File: pulse-kotlin/src/commonTest/kotlin/com/fluxchess/pulse/kotlin/engine/PositionTest.kt
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package com.fluxchess.pulse.kotlin.engine

import kotlin.test.Test
import kotlin.test.assertEquals

class PositionTest {
	@Test
	fun `Moving a white piece should set the active color to black`() {
		val position = Position()
		position.activeColor = WHITE
		position.put(WHITE_KING, E1)
		position.put(BLACK_KING, E8)
		val move = moveOf(NORMAL_MOVE, E1, E2, WHITE_KING, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(BLACK, position.activeColor)

		position.undoMove(move)

		assertEquals(WHITE, position.activeColor)
	}

	@Test
	fun `Moving a black piece should set the active color to white`() {
		val position = Position()
		position.activeColor = BLACK
		position.put(WHITE_KING, E1)
		position.put(BLACK_KING, E8)
		val move = moveOf(NORMAL_MOVE, E8, E7, BLACK_KING, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(WHITE, position.activeColor)

		position.undoMove(move)

		assertEquals(BLACK, position.activeColor)
	}

	@Test
	fun `Moving white's kingside rook should remove white's kingside castling right`() {
		val position = Position()
		position.activeColor = WHITE
		position.castlingRights = WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE
		position.put(WHITE_KING, E1)
		position.put(WHITE_ROOK, A1)
		position.put(WHITE_ROOK, H1)
		position.put(BLACK_KING, E8)
		position.put(BLACK_ROOK, A8)
		position.put(BLACK_ROOK, H8)
		val move = moveOf(NORMAL_MOVE, H1, H2, WHITE_ROOK, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE, position.castlingRights)

		position.undoMove(move)

		assertEquals(WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE, position.castlingRights)
	}

	@Test
	fun `Moving white's queenside rook should remove white's queenside castling right`() {
		val position = Position()
		position.activeColor = WHITE
		position.castlingRights = WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE
		position.put(WHITE_KING, E1)
		position.put(WHITE_ROOK, A1)
		position.put(WHITE_ROOK, H1)
		position.put(BLACK_KING, E8)
		position.put(BLACK_ROOK, A8)
		position.put(BLACK_ROOK, H8)
		val move = moveOf(NORMAL_MOVE, A1, A2, WHITE_ROOK, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(WHITE_KINGSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE, position.castlingRights)

		position.undoMove(move)

		assertEquals(WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE, position.castlingRights)
	}

	@Test
	fun `Moving white's king should remove white's kingside and queenside castling right`() {
		val position = Position()
		position.activeColor = WHITE
		position.castlingRights = WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE
		position.put(WHITE_KING, E1)
		position.put(WHITE_ROOK, A1)
		position.put(WHITE_ROOK, H1)
		position.put(BLACK_KING, E8)
		position.put(BLACK_ROOK, A8)
		position.put(BLACK_ROOK, H8)
		val move = moveOf(NORMAL_MOVE, E1, E2, WHITE_KING, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(BLACK_KINGSIDE or BLACK_QUEENSIDE, position.castlingRights)

		position.undoMove(move)

		assertEquals(WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE, position.castlingRights)
	}

	@Test
	fun `Moving black's kingside rook should remove blacks's kingside castling right`() {
		val position = Position()
		position.activeColor = BLACK
		position.castlingRights = WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE
		position.put(WHITE_KING, E1)
		position.put(WHITE_ROOK, A1)
		position.put(WHITE_ROOK, H1)
		position.put(BLACK_KING, E8)
		position.put(BLACK_ROOK, A8)
		position.put(BLACK_ROOK, H8)
		val move = moveOf(NORMAL_MOVE, H8, H7, BLACK_ROOK, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_QUEENSIDE, position.castlingRights)

		position.undoMove(move)

		assertEquals(WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE, position.castlingRights)
	}

	@Test
	fun `Moving black's queenside rook should remove blacks's queenside castling right`() {
		val position = Position()
		position.activeColor = BLACK
		position.castlingRights = WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE
		position.put(WHITE_KING, E1)
		position.put(WHITE_ROOK, A1)
		position.put(WHITE_ROOK, H1)
		position.put(BLACK_KING, E8)
		position.put(BLACK_ROOK, A8)
		position.put(BLACK_ROOK, H8)
		val move = moveOf(NORMAL_MOVE, A8, A7, BLACK_ROOK, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE, position.castlingRights)

		position.undoMove(move)

		assertEquals(WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE, position.castlingRights)
	}

	@Test
	fun `Moving black's king should remove blacks's kingside and queenside castling right`() {
		val position = Position()
		position.activeColor = BLACK
		position.castlingRights = WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE
		position.put(WHITE_KING, E1)
		position.put(WHITE_ROOK, A1)
		position.put(WHITE_ROOK, H1)
		position.put(BLACK_KING, E8)
		position.put(BLACK_ROOK, A8)
		position.put(BLACK_ROOK, H8)
		val move = moveOf(NORMAL_MOVE, E8, E7, BLACK_KING, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(WHITE_KINGSIDE or WHITE_QUEENSIDE, position.castlingRights)

		position.undoMove(move)

		assertEquals(WHITE_KINGSIDE or WHITE_QUEENSIDE or BLACK_KINGSIDE or BLACK_QUEENSIDE, position.castlingRights)
	}

	@Test
	fun `Making a pawn double move for white should set the en passant square`() {
		val position = Position()
		position.activeColor = WHITE
		position.put(WHITE_KING, E1)
		position.put(WHITE_PAWN, E2)
		position.put(BLACK_KING, E8)
		val move = moveOf(PAWN_DOUBLE_MOVE, E2, E4, WHITE_PAWN, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(E3, position.enPassantSquare)

		position.undoMove(move)

		assertEquals(NO_SQUARE, position.enPassantSquare)
	}

	@Test
	fun `Making a pawn double move for black should set the en passant square`() {
		val position = Position()
		position.activeColor = BLACK
		position.put(WHITE_KING, E1)
		position.put(BLACK_KING, E8)
		position.put(BLACK_PAWN, E7)
		val move = moveOf(PAWN_DOUBLE_MOVE, E7, E5, BLACK_PAWN, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(E6, position.enPassantSquare)

		position.undoMove(move)

		assertEquals(NO_SQUARE, position.enPassantSquare)
	}

	@Test
	fun `Making an en passant move for white should clear the en passant square`() {
		val position = Position()
		position.activeColor = WHITE
		position.enPassantSquare = E6
		position.put(WHITE_KING, E1)
		position.put(WHITE_PAWN, D5)
		position.put(BLACK_KING, E8)
		position.put(BLACK_PAWN, E5)
		val move = moveOf(EN_PASSANT_MOVE, D5, E6, WHITE_PAWN, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(NO_SQUARE, position.enPassantSquare)

		position.undoMove(move)

		assertEquals(E6, position.enPassantSquare)
	}

	@Test
	fun `Making an en passant move for black should clear the en passant square`() {
		val position = Position()
		position.activeColor = BLACK
		position.enPassantSquare = E3
		position.put(WHITE_KING, E1)
		position.put(WHITE_PAWN, E4)
		position.put(BLACK_KING, E8)
		position.put(BLACK_PAWN, D4)
		val move = moveOf(EN_PASSANT_MOVE, D4, E3, BLACK_PAWN, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(NO_SQUARE, position.enPassantSquare)

		position.undoMove(move)

		assertEquals(E3, position.enPassantSquare)
	}

	@Test
	fun `Making a move should increment the halfmove clock`() {
		val position = Position()
		position.activeColor = WHITE
		position.halfmoveClock = 0
		position.put(WHITE_KING, E1)
		position.put(BLACK_KING, E8)
		val move = moveOf(NORMAL_MOVE, E1, E2, WHITE_KING, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(1, position.halfmoveClock)

		position.undoMove(move)

		assertEquals(0, position.halfmoveClock)
	}

	@Test
	fun `Moving a pawn should reset the halfmove clock`() {
		val position = Position()
		position.activeColor = WHITE
		position.halfmoveClock = 1
		position.put(WHITE_KING, E1)
		position.put(WHITE_PAWN, E2)
		position.put(BLACK_KING, E8)
		val move = moveOf(NORMAL_MOVE, E2, E3, WHITE_PAWN, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(0, position.halfmoveClock)

		position.undoMove(move)

		assertEquals(1, position.halfmoveClock)
	}

	@Test
	fun `Capturing a piece should reset the halfmove clock`() {
		val position = Position()
		position.activeColor = WHITE
		position.halfmoveClock = 1
		position.put(WHITE_KING, E1)
		position.put(WHITE_QUEEN, D1)
		position.put(BLACK_KING, E8)
		position.put(BLACK_PAWN, D7)
		val move = moveOf(NORMAL_MOVE, D1, D7, WHITE_QUEEN, BLACK_PAWN, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(0, position.halfmoveClock)

		position.undoMove(move)

		assertEquals(1, position.halfmoveClock)
	}

	@Test
	fun `Making a move should increment the halfmove number`() {
		val position = Position()
		position.activeColor = WHITE
		position.halfmoveNumber = 0
		position.put(WHITE_KING, E1)
		position.put(BLACK_KING, E8)
		val move = moveOf(NORMAL_MOVE, E1, E2, WHITE_KING, NO_PIECE, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(1, position.halfmoveNumber)

		position.undoMove(move)

		assertEquals(0, position.halfmoveNumber)
	}

	@Test
	fun `Capturing a piece should replace the target piece by the origin piece`() {
		val position = Position()
		position.activeColor = WHITE
		position.put(WHITE_KING, E1)
		position.put(WHITE_QUEEN, D1)
		position.put(BLACK_KING, E8)
		position.put(BLACK_PAWN, D7)
		val move = moveOf(NORMAL_MOVE, D1, D7, WHITE_QUEEN, BLACK_PAWN, NO_PIECE_TYPE)

		position.makeMove(move)

		assertEquals(NO_PIECE, position.board[D1])
		assertEquals(WHITE_QUEEN, position.board[D7])

		position.undoMove(move)

		assertEquals(WHITE_QUEEN, position.board[D1])
		assertEquals(BLACK_PAWN, position.board[D7])
	}

	@Test
	fun `Making a pawn promotion move for white should replace the pawn by the promotion`() {
		val position = Position()
		position.activeColor = WHITE
		position.put(WHITE_KING, E1)
		position.put(WHITE_PAWN, C7)
		position.put(BLACK_KING, E8)
		val move = moveOf(PAWN_PROMOTION_MOVE, C7, C8, WHITE_PAWN, NO_PIECE, QUEEN)

		position.makeMove(move)

		assertEquals(NO_PIECE, position.board[C7])
		assertEquals(WHITE_QUEEN, position.board[C8])

		position.undoMove(move)

		assertEquals(WHITE_PAWN, position.board[C7])
		assertEquals(NO_PIECE, position.board[C8])
	}

	@Test
	fun `Making a pawn promotion move for black should replace the pawn by the promotion`() {
		val position = Position()
		position.activeColor = BLACK
		position.put(WHITE_KING, E1)
		position.put(BLACK_KING, E8)
		position.put(BLACK_PAWN, C2)
		val move = moveOf(PAWN_PROMOTION_MOVE, C2, C1, BLACK_PAWN, NO_PIECE, QUEEN)

		position.makeMove(move)

		assertEquals(NO_PIECE, position.board[C2])
		assertEquals(BLACK_QUEEN, position.board[C1])

		position.undoMove(move)

		assertEquals(BLACK_PAWN, position.board[C2])
		assertEquals(NO_PIECE, position.board[C1])
	}
}

File: pulse-go/internal/pulse/engine/position_test.go
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package engine

import "testing"

func TestPosition_ActiveColor(t *testing.T) {
	tests := []struct {
		name        string
		setup       func(p *Position)
		move        Move
		activeColor Color
		want        Color
	}{
		{
			name: "Moving a white piece should set the active color to black",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(BlackKing, E8)
			},
			move:        MoveOf(NormalMove, E1, E2, WhiteKing, NoPiece, NoPieceType),
			activeColor: White,
			want:        Black,
		},
		{
			name: "Moving a black piece should set the active color to white",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(BlackKing, E8)
			},
			move:        MoveOf(NormalMove, E8, E7, BlackKing, NoPiece, NoPieceType),
			activeColor: Black,
			want:        White,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			p := &Position{
				ActiveColor: tt.activeColor,
			}
			tt.setup(p)
			p.MakeMove(tt.move)
			if p.ActiveColor != tt.want {
				t.Errorf("wanted active color to be %v, but got %v", tt.want, p.ActiveColor)
			}
			p.UndoMove(tt.move)
			if p.ActiveColor != tt.activeColor {
				t.Errorf("wanted active color to be %v, but got %v", tt.activeColor, p.ActiveColor)
			}
		})
	}
}

func TestPosition_CastlingRights(t *testing.T) {
	tests := []struct {
		name           string
		setup          func(p *Position)
		move           Move
		activeColor    Color
		castlingRights Castling
		want           Castling
	}{
		{
			name: "Moving white's kingside rook should remove white's kingside castling right",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhiteRook, A1)
				p.Put(WhiteRook, H1)
				p.Put(BlackKing, E8)
				p.Put(BlackRook, A8)
				p.Put(BlackRook, H8)
			},
			move:           MoveOf(NormalMove, H1, H2, WhiteRook, NoPiece, NoPieceType),
			activeColor:    White,
			castlingRights: WhiteKingside | WhiteQueenside | BlackKingside | BlackQueenside,
			want:           WhiteQueenside | BlackKingside | BlackQueenside,
		},
		{
			name: "Moving white's queenside rook should remove white's queenside castling right",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhiteRook, A1)
				p.Put(WhiteRook, H1)
				p.Put(BlackKing, E8)
				p.Put(BlackRook, A8)
				p.Put(BlackRook, H8)
			},
			move:           MoveOf(NormalMove, A1, A2, WhiteRook, NoPiece, NoPieceType),
			activeColor:    White,
			castlingRights: WhiteKingside | WhiteQueenside | BlackKingside | BlackQueenside,
			want:           WhiteKingside | BlackKingside | BlackQueenside,
		},
		{
			name: "Moving white's king should remove white's kingside and queenside castling right",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhiteRook, A1)
				p.Put(WhiteRook, H1)
				p.Put(BlackKing, E8)
				p.Put(BlackRook, A8)
				p.Put(BlackRook, H8)
			},
			move:           MoveOf(NormalMove, E1, E2, WhiteKing, NoPiece, NoPieceType),
			activeColor:    White,
			castlingRights: WhiteKingside | WhiteQueenside | BlackKingside | BlackQueenside,
			want:           BlackKingside | BlackQueenside,
		},
		{
			name: "Moving black's kingside rook should remove black's kingside castling right",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhiteRook, A1)
				p.Put(WhiteRook, H1)
				p.Put(BlackKing, E8)
				p.Put(BlackRook, A8)
				p.Put(BlackRook, H8)
			},
			move:           MoveOf(NormalMove, H8, H7, BlackRook, NoPiece, NoPieceType),
			activeColor:    Black,
			castlingRights: WhiteKingside | WhiteQueenside | BlackKingside | BlackQueenside,
			want:           WhiteKingside | WhiteQueenside | BlackQueenside,
		},
		{
			name: "Moving black's queenside rook should remove blacks's queenside castling right",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhiteRook, A1)
				p.Put(WhiteRook, H1)
				p.Put(BlackKing, E8)
				p.Put(BlackRook, A8)
				p.Put(BlackRook, H8)
			},
			move:           MoveOf(NormalMove, A8, A7, BlackRook, NoPiece, NoPieceType),
			activeColor:    Black,
			castlingRights: WhiteKingside | WhiteQueenside | BlackKingside | BlackQueenside,
			want:           WhiteKingside | WhiteQueenside | BlackKingside,
		},
		{
			name: "Moving black's king should remove blacks's kingside and queenside castling right",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhiteRook, A1)
				p.Put(WhiteRook, H1)
				p.Put(BlackKing, E8)
				p.Put(BlackRook, A8)
				p.Put(BlackRook, H8)
			},
			move:           MoveOf(NormalMove, E8, E7, BlackKing, NoPiece, NoPieceType),
			activeColor:    Black,
			castlingRights: WhiteKingside | WhiteQueenside | BlackKingside | BlackQueenside,
			want:           WhiteKingside | WhiteQueenside,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			p := &Position{
				ActiveColor:    tt.activeColor,
				CastlingRights: tt.castlingRights,
			}
			tt.setup(p)
			p.MakeMove(tt.move)
			if p.CastlingRights != tt.want {
				t.Errorf("wanted castling rights to be %v, but got %v", tt.want, p.CastlingRights)
			}
			p.UndoMove(tt.move)
			if p.CastlingRights != tt.castlingRights {
				t.Errorf("wanted castling rights to be %v, but got %v", tt.castlingRights, p.CastlingRights)
			}
		})
	}
}

func TestPosition_EnPassantSquare(t *testing.T) {
	tests := []struct {
		name            string
		setup           func(p *Position)
		move            Move
		activeColor     Color
		enPassantSquare Square
		want            Square
	}{
		{
			name: "Making a pawn double move for white should set the en passant square",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhitePawn, E2)
				p.Put(BlackKing, E8)
			},
			move:            MoveOf(PawnDoubleMove, E2, E4, WhitePawn, NoPiece, NoPieceType),
			activeColor:     White,
			enPassantSquare: NoSquare,
			want:            E3,
		},
		{
			name: "Making a pawn double move for black should set the en passant square",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(BlackKing, E8)
				p.Put(BlackPawn, E7)
			},
			move:            MoveOf(PawnDoubleMove, E7, E5, BlackPawn, NoPiece, NoPieceType),
			activeColor:     Black,
			enPassantSquare: NoSquare,
			want:            E6,
		},
		{
			name: "Making an en passant move for white should clear the en passant square",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhitePawn, D5)
				p.Put(BlackKing, E8)
				p.Put(BlackPawn, E5)
			},
			move:            MoveOf(EnPassantMove, D5, E6, WhitePawn, NoPiece, NoPieceType),
			activeColor:     White,
			enPassantSquare: E6,
			want:            NoSquare,
		},
		{
			name: "Making an en passant move for black should clear the en passant square",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhitePawn, E4)
				p.Put(BlackKing, E8)
				p.Put(BlackPawn, D4)
			},
			move:            MoveOf(EnPassantMove, D4, E3, BlackPawn, NoPiece, NoPieceType),
			activeColor:     Black,
			enPassantSquare: E3,
			want:            NoSquare,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			p := &Position{
				ActiveColor:     tt.activeColor,
				EnPassantSquare: tt.enPassantSquare,
			}
			tt.setup(p)
			p.MakeMove(tt.move)
			if p.EnPassantSquare != tt.want {
				t.Errorf("wanted en passant square to be %v, but got %v", tt.want, p.EnPassantSquare)
			}
			p.UndoMove(tt.move)
			if p.EnPassantSquare != tt.enPassantSquare {
				t.Errorf("wanted en passant square to be %v, but got %v", tt.enPassantSquare, p.EnPassantSquare)
			}
		})
	}
}

func TestPosition_HalfmoveClock(t *testing.T) {
	tests := []struct {
		name          string
		setup         func(p *Position)
		move          Move
		activeColor   Color
		halfmoveClock int
		want          int
	}{
		{
			name: "Making a move should increment the halfmove clock",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(BlackKing, E8)
			},
			move:          MoveOf(NormalMove, E1, E2, WhiteKing, NoPiece, NoPieceType),
			activeColor:   White,
			halfmoveClock: 0,
			want:          1,
		},
		{
			name: "Moving a pawn should reset the halfmove clock",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhitePawn, E2)
				p.Put(BlackKing, E8)
			},
			move:          MoveOf(NormalMove, E2, E3, WhitePawn, NoPiece, NoPieceType),
			activeColor:   White,
			halfmoveClock: 1,
			want:          0,
		},
		{
			name: "Capturing a piece should reset the halfmove clock",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhiteQueen, D1)
				p.Put(BlackKing, E8)
				p.Put(BlackPawn, D7)
			},
			move:          MoveOf(NormalMove, D1, D7, WhiteQueen, BlackPawn, NoPieceType),
			activeColor:   White,
			halfmoveClock: 1,
			want:          0,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			p := &Position{
				ActiveColor:   tt.activeColor,
				HalfmoveClock: tt.halfmoveClock,
			}
			tt.setup(p)
			p.MakeMove(tt.move)
			if p.HalfmoveClock != tt.want {
				t.Errorf("wanted halfmove clock to be %v, but got %v", tt.want, p.HalfmoveClock)
			}
			p.UndoMove(tt.move)
			if p.HalfmoveClock != tt.halfmoveClock {
				t.Errorf("wanted halfmove clock to be %v, but got %v", tt.halfmoveClock, p.HalfmoveClock)
			}
		})
	}
}

func TestPosition_HalfmoveNumber(t *testing.T) {
	tests := []struct {
		name           string
		setup          func(p *Position)
		move           Move
		activeColor    Color
		halfmoveNumber int
		want           int
	}{
		{
			name: "Making a move should increment the halfmove number",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(BlackKing, E8)
			},
			move:           MoveOf(NormalMove, E1, E2, WhiteKing, NoPiece, NoPieceType),
			activeColor:    White,
			halfmoveNumber: 0,
			want:           1,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			p := &Position{
				ActiveColor:    tt.activeColor,
				HalfmoveNumber: tt.halfmoveNumber,
			}
			tt.setup(p)
			p.MakeMove(tt.move)
			if p.HalfmoveNumber != tt.want {
				t.Errorf("wanted halfmove number to be %v, but got %v", tt.want, p.HalfmoveNumber)
			}
			p.UndoMove(tt.move)
			if p.HalfmoveNumber != tt.halfmoveNumber {
				t.Errorf("wanted halfmove number to be %v, but got %v", tt.halfmoveNumber, p.HalfmoveNumber)
			}
		})
	}
}

func TestPosition_board(t *testing.T) {
	tests := []struct {
		name                  string
		setup                 func(p *Position)
		move                  Move
		activeColor           Color
		wantOriginSquarePiece Piece
		wantTargetSquarePiece Piece
	}{
		{
			name: "Capturing a piece should replace the target piece by the origin piece",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhiteQueen, D1)
				p.Put(BlackKing, E8)
				p.Put(BlackPawn, D7)
			},
			move:                  MoveOf(NormalMove, D1, D7, WhiteQueen, BlackPawn, NoPieceType),
			activeColor:           White,
			wantOriginSquarePiece: NoPiece,
			wantTargetSquarePiece: WhiteQueen,
		},
		{
			name: "Making a pawn promotion move for white should replace the pawn by the promotion",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(WhitePawn, C7)
				p.Put(BlackKing, E8)
			},
			move:                  MoveOf(PawnPromotionMove, C7, C8, WhitePawn, NoPiece, Queen),
			activeColor:           White,
			wantOriginSquarePiece: NoPiece,
			wantTargetSquarePiece: WhiteQueen,
		},
		{
			name: "Making a pawn promotion move for black should replace the pawn by the promotion",
			setup: func(p *Position) {
				p.Put(WhiteKing, E1)
				p.Put(BlackKing, E8)
				p.Put(BlackPawn, C2)
			},
			move:                  MoveOf(PawnPromotionMove, C2, C1, BlackPawn, NoPiece, Queen),
			activeColor:           Black,
			wantOriginSquarePiece: NoPiece,
			wantTargetSquarePiece: BlackQueen,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			p := &Position{
				ActiveColor: tt.activeColor,
			}
			tt.setup(p)
			p.MakeMove(tt.move)
			originSq := OriginSquareOf(tt.move)
			targetSq := TargetSquareOf(tt.move)
			originPc := OriginPieceOf(tt.move)
			targetPc := TargetPieceOf(tt.move)
			if p.board[originSq] != tt.wantOriginSquarePiece {
				t.Errorf("wanted piece %v to be on origin square, but got %v", tt.wantOriginSquarePiece, p.board[originSq])
			}
			if p.board[targetSq] != tt.wantTargetSquarePiece {
				t.Errorf("wanted piece %v to be on target square, but got %v", tt.wantTargetSquarePiece, p.board[targetSq])
			}
			p.UndoMove(tt.move)
			if p.board[originSq] != originPc {
				t.Errorf("wanted piece %v to be on origin square, but got %v", originPc, p.board[originSq])
			}
			if p.board[targetSq] != targetPc {
				t.Errorf("wanted piece %v to be on target square, but got %v", targetPc, p.board[targetSq])
			}
		})
	}
}
